fix(migration): migrate `deprecated_tags` to new tagging model

alexkim205 opened the pull request:
## Changes

Followup AI from https://github.com/PostHog/postmortem/pull/13

This PR aims to migrate old production tags that are stuck in `deprecated_tags` columns to the new TaggedItem-Tag system.

You'll probably want to give this ticket a 👀 to get the full context for why this migration came out the way it did. https://github.com/PostHog/posthog/issues/8771.

~24.5K of the diff is made up of snapshots.

<!-- If this is based on a reference design, include a link to the relevant Figma frame! -->

*Stay up-to-date with our [coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

## How did you test this code?

Wrote a migration test that covers these edge cases:
- Already existing tag
- Already existing taggeditem
- Duplicate tags within object
- Duplicate tags across objects within migration
- Tags from different teams
- Number of tags > `batch_size`
- Nullish tags

alexkim205 committed:
fix uniqueness

alexkim205 committed:
took out base by accident

alexkim205 committed:
revert base

alexkim205 committed:
black

alexkim205 commented on ee/test/test_migration_0012.py:
```diff
@@ -0,0 +1,94 @@
+        # TaggedItems = team2_total_property_definitions * 3 + team1_taggeditems
+        self.assertEqual(Tag.objects.filter(team_id=self.team.id).all().count(), 5)
+        self.assertEqual(Tag.objects.all().count(), self.team2_total_property_definitions + 2 + 5)
+        self.assertEqual(TaggedItem.objects.all().count(), self.team2_total_property_definitions * 3 + 7)
```
It would be the icing on the cake if we could also assert on the number of transactions that have occurred, but adding boilerplate to support that seems like scope creep. Any suggestions on how to do it easily are welcome!

alexkim205 committed:
remove check

macobo reviewed (commented):
Only reviewing this from a code/migration issues standpoint, didn't take a look at correctness.

There's nothing immediately suspect here besides the try-catch.

I suggest running the loops (without actually creating anything) once against production as well as that will probably reveal issues. https://posthog.com/handbook/engineering/production-access

macobo commented on ee/migrations/0012_migrate_tags_v2.py:
```diff
@@ -0,0 +1,83 @@
+
+    try:
+        # Collect insight tags and taggeditems
+        for [tags, team_id, id] in EnterpriseEventDefinition.objects.exclude(
```
Nit: This is unpythonic. Either do `(tags, team_id, id)` or `tags, team_id, id`

macobo commented on ee/migrations/0012_migrate_tags_v2.py:
```diff
@@ -0,0 +1,83 @@
+        for [tags, team_id, id] in EnterprisePropertyDefinition.objects.exclude(
+            Q(deprecated_tags__isnull=True) | Q(deprecated_tags=[]),
+        ).values_list("deprecated_tags", "team_id", "id"):
+            unique_tags = set([tagify(t) for t in tags if isinstance(t, str) and t.strip() != ""])
```
Nit: Remove `[`

guidoiaquinti commented on posthog/migrations/0217_migrate_tags_v2.py:
```diff
@@ -0,0 +1,82 @@
+        # Create tag <-> item relationships, ignoring conflicts
+        TaggedItem.objects.bulk_create([pair[1] for pair in createables], ignore_conflicts=True, batch_size=batch_size)
+
+    except Exception as e:
```
Ditto

macobo commented on ee/test/test_migration_0012.py:
```diff
@@ -0,0 +1,103 @@
+from posthog.test.base import TestMigrations
+
+
+class TagsTestCase(TestMigrations):
```
If you're adding tests I suggest using the snapshot functions here as well to document the generated SQL

guidoiaquinti commented on posthog/test/base.py:
```diff
@@ -1,4 +1,5 @@
 import inspect
+import logging
```
Is this needed?

guidoiaquinti commented on posthog/utils.py:
```diff
@@ -5,6 +5,7 @@
 import gzip
 import hashlib
 import json
+import logging
```
Can we use `structlog` as per [docs](https://posthog.com/docs/contribute/coding-conventions#format) ? 🙏

alexkim205 replied:
Ah thank you for the link!

macobo commented on ee/migrations/0012_migrate_tags_v2.py:
```diff
@@ -0,0 +1,83 @@
+
+            # Tags that are in `tags_to_create` but not in `created_tags` are tags that already exist
+            # in the db and must be fetched individually.
+            for pair in createables[i : (i + batch_size)]:
```
Nit: `for tag, tagged_item in `

hazzadous commented on ee/migrations/0012_migrate_tags_v2.py:
```diff
@@ -0,0 +1,83 @@
+        # Attempts to create tags in bulk while ignoring conflicts. bulk_create does not return any data
+        # about which tags were ignored and created, so we must take care of this manually.
+        tags_to_create = [pair[0] for pair in createables]
+        Tag.objects.bulk_create(tags_to_create, ignore_conflicts=True, batch_size=batch_size)
```
Note that this will try to insert a tag for each tag item. I don't think it will be an issue but it's more than we need.

alexkim205 replied:
It's suboptimal and I would've preferred to do an in_bulk check that can fetch all tags that already exist before bulk creating them like this. I couldn't think of a way to do this using django's limited API here, so just opted for the simplest solution here.

hazzadous commented on ee/migrations/0012_migrate_tags_v2.py:
```diff
@@ -0,0 +1,83 @@
+
+        # Attempts to create tags in bulk while ignoring conflicts. bulk_create does not return any data
+        # about which tags were ignored and created, so we must take care of this manually.
+        tags_to_create = [pair[0] for pair in createables]
```
Rather than using just `pair` can we unpack into meaningfully named vars, and use `_` if we do not need to reference, e.g.

```
tags_to_create = [tag for (tag, _) in creatables]
```

hazzadous commented on ee/migrations/0012_migrate_tags_v2.py:
```diff
@@ -0,0 +1,83 @@
+            batch = tags_to_create[i : (i + batch_size)]
+
+            # Find tags that were created, and not already existing
+            created_tags = Tag.objects.in_bulk([t.id for t in batch])
```
If bulk create doesn't return anything, won't this t.id always be `None`?

Also, I don't think we need two layers of batching here, but batch up the taggeditems, for each batch fetch the tag_id associated with (name, team_id), bulk insert that batch.

alexkim205 replied:
> If bulk create doesn't return anything, won't this t.id always be None?

Good q, if `bulk_create` doesn't create any new tags, `in_bulk` will return an empty list which indicates that all the tags to be created already existed. I don't think `t` here will ever be an empty object since any object returned here must already exist in the table.

> Also, I don't think we need two layers of batching here, but batch up the taggeditems, for each batch fetch the tag_id associated with (name, team_id), bulk insert that batch.

👍 makes sense!

hazzadous commented on ee/test/test_migration_0012.py:
```diff
@@ -0,0 +1,103 @@
+            property_definition_id=EnterprisePropertyDefinition.objects.filter(team_id=self.team2.id).first().id,
+        )
+
+    def test_tags_migrated(self):
```
Can we run the migration twice?

non-blocking: I'm not sure how that fits in with the `TestMigration` structure, but tbh I'd prefer to just perform the migrations explicitly in these tests rather implicitly somewhere in the class hierarchy

alexkim205 replied:
I think this will require some custom boilerplate since `TestMigration` intercepts the one time `setUp` call to perform the migration. I agree it'd be nice to have separate setup, execute, and teardown methods for each test but I think it'll be fine for now.

hazzadous commented on posthog/migrations/0217_migrate_tags_v2.py:
```diff
@@ -0,0 +1,82 @@
+# Generated by Django 3.2.5 on 2022-03-01 23:41
```
What's with the two migrations? Is this just because they are in diff django apps?

alexkim205 replied:
It's a weird consequence/code smell of splitting out event_definitions and property_definitions into a seperate enterprise app, which requires them to be migrated separately.

alexkim205 committed:
create indices concurrently and add a test

alexkim205 committed:
Merge remote-tracking branch 'origin' into chore/fix-uniqueness-constraint-tagged-items

alexkim205 commented on ee/migrations/0012_migrate_tags_v2.py:
```diff
@@ -0,0 +1,83 @@
+        # Create tag <-> item relationships, ignoring conflicts
+        TaggedItem.objects.bulk_create([pair[1] for pair in createables], ignore_conflicts=True, batch_size=batch_size)
+
+    except Exception as e:
```
Yeah that makes a lot of sense. I think the issue with the first version of the migration was that Django wasn't logging any errors which made it difficult to debug what was going on, so I just threw in a try catch that would intercept the error, print it, and then throw it again (which I forgot to do 😅).

There isn't a specific error that I wanted to recover from here and adding logging info like you suggested below will give us some clue where an error happens so I think it makes sense to remove this try catch entirely.

alexkim205 commented:
Appreciate the reviews! Addressed all feedback.

@macobo Unfortunately it looks like I can't run the read-only part of the migration on production data, since I'm unable to access the deprecated `deprecated_tags` field from the Django shell. The `django_deprecate_fields` library only allows accessing this field in migration related commands, which makes sense but doesn't work for in this case. 

```python
// site-packages/django_deprecate_fields/deprecate_field.py
...
if not set(sys.argv) & {"makemigrations", "migrate", "showmigrations"}:
    return DeprecatedField(return_instead)
...
```

The next closest thing that came to mind was to run the SQL queries that are generated by the migration on Metabase. All queries didn't require any batching since the total counts in production were all well under the batch_size (1000).

```
  SELECT COUNT(*) AS "__count"
  FROM "posthog_dashboarditem"
  WHERE NOT (("posthog_dashboarditem"."deprecated_tags" IS NULL
              OR ("posthog_dashboarditem"."deprecated_tags" = ARRAY[]::varchar(32)[]
                  AND "posthog_dashboarditem"."deprecated_tags" IS NOT NULL)))

 > 113
```

```
  SELECT COUNT(*) AS "__count"
  FROM "posthog_dashboard"
  WHERE NOT (("posthog_dashboard"."deprecated_tags" IS NULL
              OR ("posthog_dashboard"."deprecated_tags" = ARRAY[]::varchar(32)[]
                  AND "posthog_dashboard"."deprecated_tags" IS NOT NULL)))

 > 78
```

```
  SELECT COUNT(*) AS "__count"
  FROM "ee_enterpriseeventdefinition"
  INNER JOIN "posthog_eventdefinition" ON ("ee_enterpriseeventdefinition"."eventdefinition_ptr_id" = "posthog_eventdefinition"."id")
  WHERE NOT (("ee_enterpriseeventdefinition"."deprecated_tags" IS NULL
              OR ("ee_enterpriseeventdefinition"."deprecated_tags" = ARRAY[]::varchar(32)[]
                  AND "ee_enterpriseeventdefinition"."deprecated_tags" IS NOT NULL)))

 > 10
```

```
  SELECT COUNT(*) AS "__count"
  FROM "ee_enterprisepropertydefinition"
  INNER JOIN "posthog_propertydefinition" ON ("ee_enterprisepropertydefinition"."propertydefinition_ptr_id" = "posthog_propertydefinition"."id")
  WHERE NOT (("ee_enterprisepropertydefinition"."deprecated_tags" IS NULL
              OR ("ee_enterprisepropertydefinition"."deprecated_tags" = ARRAY[]::varchar(32)[]
                  AND "ee_enterprisepropertydefinition"."deprecated_tags" IS NOT NULL)))

 > 0
```

alexkim205 commented:
Also worth mentioning that the large number of one-off queries in the snapshot represent the absolute worst case scenario where 1001 objects have overlapping tags. A quick query in metabase tells me that's not going to be a problem in practice.

alexkim205 committed:
use not null

alexkim205 committed:
Merge remote-tracking branch 'origin' into chore/fix-uniqueness-constraint-tagged-items

alexkim205 committed:
dedupe as first operation

alexkim205 committed:
rebased

alexkim205 committed:
non deterministic tests

alexkim205 committed:
ws

alexkim205 committed:
update snappys

hazzadous reviewed (approved):
Looks good

🤞 the tests cover all the cases we need. It's a shame about the duplication, we seem to be a prisoner of django app structure which I think we could probably ignore to some extent (i.e. I think we could actually put this into one data migration in either ee or posthog).

Lots of yak shaving to do, happy to get in

alexkim205 committed:
Merge remote-tracking branch 'origin' into chore/fix-uniqueness-constraint-tagged-items

alexkim205 committed:
Merge remote-tracking branch 'origin' into fix/tags-migration-v2

alexkim205 committed:
ws

alexkim205 committed:
ws
